Lib/asyncio/base_events.py

```diff
@@ -452,6 +452,11 @@ def run_until_complete(self, future):
         try:
             self.run_forever()
         except:
+            if new_task and not future.done():
+                # A signal from outside interrupted the execution. Cancel the
+                # coroutine and wait for the exit handlers to finish
+                future.cancel()
+                self.run_forever()
```

Comment:
What's a 'signal'?  It could be a simple unhandled exception. Right now, the 'except' clause is too broad to do what you are doing here.

Reply:
If we are running a coroutine (ie `new_task is True`), and `future.done()` is false, then the exception is not coming from inside the coroutine and has to be some kind of external signal.

The raw `except:` clause is very broad indeed, but the exception can actually be any kind of user defined exception here, so it makes sense to catch them all. For instance, the signal example from the docs would raise an OSError:

```python
import asyncio, signal

def handler(signum, frame):
    raise OSError("Couldn't open device!")
signal.signal(signal.SIGALRM, handler)
signal.alarm(1)

loop = asyncio.get_event_loop()
loop.run_until_complete(asyncio.sleep(10))
```

And in this example, the `sleep(10)` coroutine is never cancelled. This isn't an issue here, but is an issue in the example I provided with the PR: if you mix async contextes into this, the `__aexit__` are not called